tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_replay_cache.c

```diff
@@ -0,0 +1,203 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include "mesh_test.h"
+#include "settings_test_backend.h"
+#include "mesh/mesh.h"
+#include "mesh/net.h"
+
+#define LOG_MODULE_NAME test_rpc
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(LOG_MODULE_NAME, LOG_LEVEL_INF);
+
+#define WAIT_TIME 60 /*seconds*/
+#define TEST_DATA_WAITING_TIME 5 /* seconds */
+
+static const struct bt_mesh_test_cfg tx_cfg = {
+	.addr = 0x0001,
+	.dev_key = { 0x01 },
+};
+static const struct bt_mesh_test_cfg rx_cfg = {
+	.addr = 0x0002,
+	.dev_key = { 0x02 },
+};
+
+static bool are_settings_kept;
+static bool is_tx_succeeded;
+
+static void test_tx_init(void)
+{
+	bt_mesh_test_cfg_set(&tx_cfg, WAIT_TIME);
+}
+
+static void test_rx_init(void)
+{
+	bt_mesh_test_cfg_set(&rx_cfg, WAIT_TIME);
+}
+
+static void test_delete(void)
+{
+	if (!are_settings_kept) {
+		settings_test_backend_clear();
+	}
+}
+
+static void tx_started(uint16_t dur, int err, void *data)
+{
+	if (err) {
+		FAIL("Couldn't start sending (err: %d)", err);
+	}
+
+	LOG_INF("Sending started");
+}
+
+static void tx_ended(int err, void *data)
+{
+	struct k_sem *sem = data;
+
+	if (err) {
+		is_tx_succeeded = false;
+		LOG_INF("Sending failed (%d)", err);
+	} else {
+		is_tx_succeeded = true;
+		LOG_INF("Sending succeeded");
+	}
+
+	k_sem_give(sem);
+}
+
+static void test_tx_immediate_replay_attack(void)
```

Comment:
Function `test_delete` will be called in any case whether a test has been failed or not. (Except fault cases like segmentation fault).

Reply:
```
Starting rpc_tx_immediate_replay_attack as device #0
Starting rpc_rx_immediate_replay_attack as device #1
Starting phy with 2 devices
d_01: @00:00:00.000000  *** Booting Zephyr OS build zephyr-v2.6.0-2267-g8ffb682b71e2  ***
d_01: @00:00:00.000000  [00:00:00.000,000] <inf> settings_test_backend: file path: mesh_replay_attack_1.log
d_00: @00:00:00.000000  *** Booting Zephyr OS build zephyr-v2.6.0-2267-g8ffb682b71e2  ***
d_00: @00:00:00.000000  [00:00:00.000,000] <inf> settings_test_backend: file path: mesh_replay_attack_0.log
d_01: @00:00:00.000000  [00:00:00.000,000] <inf> bt_hci_core: HW Platform: Nordic Semiconductor (0x0002)
d_01: @00:00:00.000000  [00:00:00.000,000] <inf> bt_hci_core: HW Variant: nRF52x (0x0002)
d_01: @00:00:00.000000  [00:00:00.000,000] <inf> bt_hci_core: Firmware: Standard Bluetooth controller (0x00) Version 2.6 Build 99
d_01: @00:00:00.000000  [00:00:00.000,000] <inf> bt_hci_core: No ID address. App must call settings_load()
d_01: @00:00:00.000000  [00:00:00.000,000] <inf> mesh_test: Bluetooth initialized
d_01: @00:00:00.000000  [00:00:00.000,000] <inf> mesh_test: Loading stored settings
d_01: @00:00:00.000000  [00:00:00.000,000] <wrn> bt_id: No static addresses stored in controller
d_01: @00:00:00.002648  [00:00:00.002,624] <inf> bt_hci_core: Identity: ED:71:8F:C2:E4:6E (random)
d_01: @00:00:00.002648  [00:00:00.002,624] <inf> bt_hci_core: HCI: version 5.2 (0x0b) revision 0x0000, manufacturer 0x0059
d_01: @00:00:00.002648  [00:00:00.002,624] <inf> bt_hci_core: LMP: version 5.2 (0x0b) subver 0xffff
d_01: @00:00:00.002648  [00:00:00.002,624] <inf> bt_mesh_main: Primary Element: 0x0002
d_01: @00:00:00.002648  [00:00:00.002,624] <dbg> bt_mesh_main.bt_mesh_provision: net_idx 0x0000 flags 0x00 iv_index 0x0000
d_01: @00:00:00.002648  [00:00:00.002,624] <dbg> bt_mesh_settings.bt_mesh_settings_store_schedule: Waiting 2000 ms vs rem 0 ms
d_01: @00:00:00.002648  [00:00:00.002,624] <dbg> bt_mesh_settings.bt_mesh_settings_store_schedule: Waiting 0 ms vs rem 2010 ms
d_01: @00:00:00.002648  [00:00:00.002,624] <dbg> bt_mesh_settings.bt_mesh_settings_store_schedule: Waiting 0 ms vs rem 0 ms
d_01: @00:00:00.002648  [00:00:00.002,624] <dbg> bt_mesh_settings.store_pending: 
d_00: @00:00:00.000000  [00:00:00.000,000] <inf> bt_hci_core: HW Platform: Nordic Semiconductor (0x0002)
d_00: @00:00:00.000000  [00:00:00.000,000] <inf> bt_hci_core: HW Variant: nRF52x (0x0002)
d_00: @00:00:00.000000  [00:00:00.000,000] <inf> bt_hci_core: Firmware: Standard Bluetooth controller (0x00) Version 2.6 Build 99
d_00: @00:00:00.000000  [00:00:00.000,000] <inf> bt_hci_core: No ID address. App must call settings_load()
d_00: @00:00:00.000000  [00:00:00.000,000] <inf> mesh_test: Bluetooth initialized
d_00: @00:00:00.000000  [00:00:00.000,000] <inf> mesh_test: Loading stored settings
d_00: @00:00:00.000000  [00:00:00.000,000] <wrn> bt_id: No static addresses stored in controller
d_00: @00:00:00.002648  [00:00:00.002,624] <inf> bt_hci_core: Identity: ED:3B:20:15:18:12 (random)
d_00: @00:00:00.002648  [00:00:00.002,624] <inf> bt_hci_core: HCI: version 5.2 (0x0b) revision 0x0000, manufacturer 0x0059
d_00: @00:00:00.002648  [00:00:00.002,624] <inf> bt_hci_core: LMP: version 5.2 (0x0b) subver 0xffff
d_00: @00:00:00.002648  [00:00:00.002,624] <inf> bt_mesh_main: Primary Element: 0x0001
d_00: @00:00:00.002648  [00:00:00.002,624] <dbg> bt_mesh_main.bt_mesh_provision: net_idx 0x0000 flags 0x00 iv_index 0x0000
d_00: @00:00:00.002648  [00:00:00.002,624] <dbg> bt_mesh_settings.bt_mesh_settings_store_schedule: Waiting 2000 ms vs rem 0 ms
d_00: @00:00:00.002648  [00:00:00.002,624] <dbg> bt_mesh_settings.bt_mesh_settings_store_schedule: Waiting 0 ms vs rem 2010 ms
d_00: @00:00:00.002648  [00:00:00.002,624] <dbg> bt_mesh_settings.bt_mesh_settings_store_schedule: Waiting 0 ms vs rem 0 ms
d_00: @00:00:00.002648  [00:00:00.002,624] <dbg> bt_mesh_settings.store_pending: 
d_01: @00:00:00.002648  [00:00:00.002,624] <inf> mesh_test: Mesh initialized
d_01: @00:00:00.002768  [00:00:00.002,746] <dbg> bt_mesh_settings.bt_mesh_settings_store_schedule: Waiting 2000 ms vs rem 0 ms
d_01: @00:00:00.002928  [00:00:00.002,899] <dbg> bt_mesh_settings.bt_mesh_settings_store_schedule: Waiting 2000 ms vs rem 2010 ms
d_00: @00:00:00.002648  [00:00:00.002,624] <inf> mesh_test: Mesh initialized
d_00: @00:00:00.002768  [00:00:00.002,746] <dbg> bt_mesh_settings.bt_mesh_settings_store_schedule: Waiting 2000 ms vs rem 0 ms
d_00: @00:00:00.002928  [00:00:00.002,899] <dbg> bt_mesh_settings.bt_mesh_settings_store_schedule: Waiting 2000 ms vs rem 2010 ms
d_00: @00:00:00.004568  [00:00:00.004,547] <inf> test_rpc: Sending started
d_01: @00:00:00.246241  [00:00:00.246,215] <dbg> bt_mesh_settings.bt_mesh_settings_store_schedule: Waiting 2000 ms vs rem 1770 ms
d_01: @00:00:00.246351  [00:00:00.246,337] <inf> mesh_test: 	len: 20 bytes
d_01: @00:00:00.246351  [00:00:00.246,337] <inf> mesh_test: 	src: 0x0001
d_01: @00:00:00.246351  [00:00:00.246,337] <inf> mesh_test: 	dst: 0x0002
d_01: @00:00:00.246351  [00:00:00.246,337] <inf> mesh_test: 	ttl: 7
d_01: @00:00:00.246351  [00:00:00.246,337] <inf> mesh_test: 	rssi: -35
d_01: @00:00:00.246351  [00:00:00.246,337] <inf> test_rpc: Receiving succeeded
d_00: @00:00:00.252284  [00:00:00.252,258] <dbg> bt_mesh_settings.bt_mesh_settings_store_schedule: Waiting 2000 ms vs rem 1760 ms
d_00: @00:00:00.252284  [00:00:00.252,258] <inf> test_rpc: Sending succeeded
d_00: @00:00:00.359981  [00:00:00.359,954] <inf> test_rpc: Sending started
d_01: @00:00:00.610742  [00:00:00.610,717] <dbg> bt_mesh_settings.bt_mesh_settings_store_schedule: Waiting 2000 ms vs rem 1400 ms
d_01: @00:00:00.610852  [00:00:00.610,839] <inf> mesh_test: 	len: 20 bytes
d_01: @00:00:00.610852  [00:00:00.610,839] <inf> mesh_test: 	src: 0x0001
d_01: @00:00:00.610852  [00:00:00.610,839] <inf> mesh_test: 	dst: 0x0002
d_01: @00:00:00.610852  [00:00:00.610,839] <inf> mesh_test: 	ttl: 7
d_01: @00:00:00.610852  [00:00:00.610,839] <inf> mesh_test: 	rssi: -35
d_01: @00:00:00.610852  [00:00:00.610,839] <inf> test_rpc: Receiving succeeded
d_00: @00:00:00.643839  [00:00:00.643,829] <dbg> bt_mesh_settings.bt_mesh_settings_store_schedule: Waiting 2000 ms vs rem 1370 ms
d_00: @00:00:00.643839  [00:00:00.643,829] <inf> test_rpc: Sending succeeded
d_00: @00:00:00.719234  [00:00:00.719,207] <inf> test_rpc: Sending started
d_01: @00:00:00.969506  [00:00:00.969,482] <dbg> bt_mesh_settings.bt_mesh_settings_store_schedule: Waiting 2000 ms vs rem 1040 ms
d_01: @00:00:00.969616  [00:00:00.969,604] <inf> mesh_test: 	len: 20 bytes
d_01: @00:00:00.969616  [00:00:00.969,604] <inf> mesh_test: 	src: 0x0001
d_01: @00:00:00.969616  [00:00:00.969,604] <inf> mesh_test: 	dst: 0x0002
d_01: @00:00:00.969616  [00:00:00.969,604] <inf> mesh_test: 	ttl: 7
d_01: @00:00:00.969616  [00:00:00.969,604] <inf> mesh_test: 	rssi: -35
d_01: @00:00:00.969616  [00:00:00.969,604] <inf> test_rpc: Receiving succeeded
d_00: @00:00:00.978785  [00:00:00.978,759] <dbg> bt_mesh_settings.bt_mesh_settings_store_schedule: Waiting 2000 ms vs rem 1030 ms
d_00: @00:00:00.978785  [00:00:00.978,759] <inf> test_rpc: Sending succeeded
d_00: @00:00:01.078487  [00:00:01.078,460] <inf> test_rpc: Sending started
d_00: @00:00:02.005828  [00:00:02.005,828] <dbg> bt_mesh_settings.store_pending: 
d_01: @00:00:02.005828  [00:00:02.005,828] <dbg> bt_mesh_settings.store_pending: 
d_00: @00:00:02.075683  [00:00:02.075,683] <err> bt_mesh_transport: Ran out of retransmit attempts
d_00: @00:00:02.075683  [00:00:02.075,683] <inf> test_rpc: Sending failed (-110)
d_00: @00:00:02.076542  [00:00:02.076,538] <inf> test_rpc: Sending started
d_00: @00:00:03.193359  [00:00:03.193,359] <err> bt_mesh_transport: Ran out of retransmit attempts
d_00: @00:00:03.193359  [00:00:03.193,359] <inf> test_rpc: Sending failed (-110)
d_00: @00:00:03.194218  [00:00:03.194,213] <inf> test_rpc: Sending started
d_00: @00:00:04.311035  [00:00:04.311,035] <err> bt_mesh_transport: Ran out of retransmit attempts
d_00: @00:00:04.311035  [00:00:04.311,035] <inf> test_rpc: Sending failed (-110)
d_00: @00:00:04.311035 INFO: test_tx_immediate_replay_attack PASSED
d_01: @00:00:29.947723 INFO: test_rx_immediate_replay_attack PASSED
d_00: @00:01:00.000000  [00:01:00.000,000] <inf> test_rpc: ************************************
d_01: @00:00:59.998546  [00:00:59.998,535] <inf> test_rpc: ************************************
Starting rpc_tx_power_replay_attack as device #0
Starting rpc_rx_power_replay_attack as device #1
Starting phy with 2 devices
d_01: @00:00:00.000000  *** Booting Zephyr OS build zephyr-v2.6.0-2267-g8ffb682b71e2  ***
d_01: @00:00:00.000000  [00:00:00.000,000] <inf> settings_test_backend: file path: mesh_replay_attack_1.log
d_00: @00:00:00.000000  *** Booting Zephyr OS build zephyr-v2.6.0-2267-g8ffb682b71e2  ***
d_00: @00:00:00.000000  [00:00:00.000,000] <inf> settings_test_backend: file path: mesh_replay_attack_0.log
d_01: @00:00:00.000000  [00:00:00.000,000] <inf> bt_hci_core: HW Platform: Nordic Semiconductor (0x0002)
d_01: @00:00:00.000000  [00:00:00.000,000] <inf> bt_hci_core: HW Variant: nRF52x (0x0002)
d_01: @00:00:00.000000  [00:00:00.000,000] <inf> bt_hci_core: Firmware: Standard Bluetooth controller (0x00) Version 2.6 Build 99
d_01: @00:00:00.000000  [00:00:00.000,000] <inf> bt_hci_core: No ID address. App must call settings_load()
d_01: @00:00:00.000000  [00:00:00.000,000] <inf> mesh_test: Bluetooth initialized
d_01: @00:00:00.000000  [00:00:00.000,000] <inf> mesh_test: Loading stored settings
d_01: @00:00:00.000000  [00:00:00.000,000] <inf> settings_test_backend: loading entry: bt/hash=4d9cb8567f0115cc8403a06aae24d802
d_01: @00:00:00.000000  
d_01: @00:00:00.000000  [00:00:00.000,000] <inf> settings_test_backend: loading entry: bt/mesh/Net=020002000000000000000000000000000000
d_01: @00:00:00.000000  
d_01: @00:00:00.000000  [00:00:00.000,000] <dbg> bt_mesh_settings.val
d_01: @00:00:00.000000                                            02 00 02 00 00 00 00 00  00 00 00 00 00 00 00 00 |........ ........
d_01: @00:00:00.000000                                            00 00                                            |..               
d_01: @00:00:00.000000  [00:00:00.000,000] <inf> settings_test_backend: loading entry: bt/mesh/NetKey/0=000102030000000000000000000000000000000000000000000000000000000000
d_01: @00:00:00.000000  
d_01: @00:00:00.000000  [00:00:00.000,000] <dbg> bt_mesh_settings.val
d_01: @00:00:00.000000                                            00 01 02 03 00 00 00 00  00 00 00 00 00 00 00 00 |........ ........
d_01: @00:00:00.000000                                            00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 |........ ........
d_01: @00:00:00.000000                                            00                                               |.                
d_01: @00:00:00.000000  [00:00:00.000,000] <inf> settings_test_backend: loading entry: bt/mesh/IV=0000000000
d_01: @00:00:00.000000  
d_01: @00:00:00.000000  [00:00:00.000,000] <dbg> bt_mesh_settings.val
d_01: @00:00:00.000000                                            00 00 00 00 00                                   |.....            
d_01: @00:00:00.000000  [00:00:00.000,000] <inf> settings_test_backend: loading entry: bt/mesh/Seq=000000
d_01: @00:00:00.000000  
d_01: @00:00:00.000000  [00:00:00.000,000] <dbg> bt_mesh_settings.val
d_01: @00:00:00.000000                                            00 00 00                                         |...              
d_01: @00:00:00.000000  [00:00:00.000,000] <inf> settings_test_backend: loading entry: bt/mesh/RPL/1=10000056
d_01: @00:00:00.000000  
d_01: @00:00:00.000000  [00:00:00.000,000] <dbg> bt_mesh_settings.val
d_01: @00:00:00.000000                                            10 00 00 56                                      |...V             
d_01: @00:00:00.000000  [00:00:00.000,000] <inf> settings_test_backend: loading entry: bt/mesh/AppKey/0=0000000405060000000000000000000000000000000000000000000000000000000000
d_01: @00:00:00.000000  
d_01: @00:00:00.000000  [00:00:00.000,000] <dbg> bt_mesh_settings.val
d_01: @00:00:00.000000                                            00 00 00 04 05 06 00 00  00 00 00 00 00 00 00 00 |........ ........
d_01: @00:00:00.000000                                            00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 |........ ........
d_01: @00:00:00.000000                                            00 00 00                                         |...              
d_01: @00:00:00.000000  [00:00:00.000,000] <inf> settings_test_backend: loading entry: bt/mesh/s/2/bind=0000
d_01: @00:00:00.000000  
d_01: @00:00:00.000000  [00:00:00.000,000] <wrn> bt_id: No static addresses stored in controller
d_01: @00:00:00.002648  [00:00:00.002,624] <inf> bt_hci_core: Identity: ED:71:8F:C2:E4:6E (random)
d_01: @00:00:00.002648  [00:00:00.002,624] <inf> bt_hci_core: HCI: version 5.2 (0x0b) revision 0x0000, manufacturer 0x0059
d_01: @00:00:00.002648  [00:00:00.002,624] <inf> bt_hci_core: LMP: version 5.2 (0x0b) subver 0xffff
d_00: @00:00:00.000000  [00:00:00.000,000] <inf> bt_hci_core: HW Platform: Nordic Semiconductor (0x0002)
d_00: @00:00:00.000000  [00:00:00.000,000] <inf> bt_hci_core: HW Variant: nRF52x (0x0002)
d_00: @00:00:00.000000  [00:00:00.000,000] <inf> bt_hci_core: Firmware: Standard Bluetooth controller (0x00) Version 2.6 Build 99
d_00: @00:00:00.000000  [00:00:00.000,000] <inf> bt_hci_core: No ID address. App must call settings_load()
d_00: @00:00:00.000000  [00:00:00.000,000] <inf> mesh_test: Bluetooth initialized
d_00: @00:00:00.000000  [00:00:00.000,000] <inf> mesh_test: Loading stored settings
d_00: @00:00:00.000000  [00:00:00.000,000] <wrn> bt_id: No static addresses stored in controller
d_00: @00:00:00.002648  [00:00:00.002,624] <inf> bt_hci_core: Identity: ED:3B:20:15:18:12 (random)
d_00: @00:00:00.002648  [00:00:00.002,624] <inf> bt_hci_core: HCI: version 5.2 (0x0b) revision 0x0000, manufacturer 0x0059
d_00: @00:00:00.002648  [00:00:00.002,624] <inf> bt_hci_core: LMP: version 5.2 (0x0b) subver 0xffff
d_00: @00:00:00.002648  [00:00:00.002,624] <inf> bt_mesh_main: Primary Element: 0x0001
d_00: @00:00:00.002648  [00:00:00.002,624] <dbg> bt_mesh_main.bt_mesh_provision: net_idx 0x0000 flags 0x00 iv_index 0x0000
d_00: @00:00:00.002648  [00:00:00.002,624] <dbg> bt_mesh_settings.bt_mesh_settings_store_schedule: Waiting 2000 ms vs rem 0 ms
d_00: @00:00:00.002648  [00:00:00.002,624] <dbg> bt_mesh_settings.bt_mesh_settings_store_schedule: Waiting 0 ms vs rem 2010 ms
d_00: @00:00:00.002648  [00:00:00.002,624] <dbg> bt_mesh_settings.bt_mesh_settings_store_schedule: Waiting 0 ms vs rem 0 ms
d_00: @00:00:00.002648  [00:00:00.002,624] <dbg> bt_mesh_settings.store_pending: 
d_01: @00:00:00.003848  [00:00:00.003,845] <wrn> bt_hci_core: opcode 0x2005 status 0x0c
d_01: @00:00:00.003848  [00:00:00.003,845] <wrn> bt_id: Ignoring failure to set address for passive scan (-5)
d_01: @00:00:00.003848  [00:00:00.003,845] <inf> bt_mesh_main: Primary Element: 0x0002
d_01: @00:00:00.003848  [00:00:00.003,845] <dbg> bt_mesh_main.bt_mesh_provision: net_idx 0x0000 flags 0x00 iv_index 0x0000
d_01: @00:00:00.003848  [00:00:00.003,845] <inf> mesh_test: Using stored settings
d_01: @00:00:00.003848 ERROR: (WEST_TOPDIR/zephyr/tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_replay_cache.c:198): 
d_01: @00:00:00.003848  [00:00:00.003,845] <inf> test_rpc: ************************************
d_01: @00:00:00.003848  main: The TESTCASE FAILED with return code 2
d_00: @00:00:00.002648  [00:00:00.002,624] <inf> mesh_test: Mesh initialized
d_00: @00:00:00.002768  [00:00:00.002,746] <dbg> bt_mesh_settings.bt_mesh_settings_store_schedule: Waiting 2000 ms vs rem 0 ms
d_00: @00:00:00.002928  [00:00:00.002,899] <dbg> bt_mesh_settings.bt_mesh_settings_store_schedule: Waiting 2000 ms vs rem 2010 ms
d_00: @00:00:00.003168 ERROR: (WEST_TOPDIR/zephyr/tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_replay_cache.c:153): 
d_00: @00:00:00.003168  [00:00:00.003,143] <inf> test_rpc: ************************************
d_00: @00:00:00.003168  main: The TESTCASE FAILED with return code 2
```
Sorry for the huge log but I added `FAIL("")` as you recommended and added the stars string into `test_delete` function. Looks like it is always called. Do I need more changes to break it?  
I'd prefer to keep this approach since it is based on BabbleSim API.

Reply:
Checked on a couple of PCs. Seems `test_delete` is really not called on part of them. BabbleSim bug? 
Changed file cleaning in @PavelVPV's suggested manner.